### ISSUE TITLE
Killing cl.exe yields unhandled error code 1

### ISSUE THREAD

**Issue by aganea** [fastbuild/fastbuild#411]
We've discovered something interesting on our build system: occasionaly builds were failing, and we didn't understand why. It was always MSVC compilation (cl.exe) which was failing remotely, on a machine belonging to a real user (not a build machine, we have both mixed which participate to the build).

We ended investigating each case with the user who was logged on the filing PC. It turned out that FBWorker.exe was slowing down their computer and that they manually killed cl.exe.

Our build system indeed reports the agent's result, with an error code:

` Failed to build Object (error 0x1) 'C:\fastbuild-work\.fbuild.tmp\0x00000000\core_1005\myproject_unity15.obj'`

Of interest here is the error code (0x1). It turns out, from the stack overflow post below, that error code 1 means that the process was terminated by the Task Manager:

https://stackoverflow.com/questions/4344923/process-exit-code-when-process-is-killed-forcibly

I was thusly wondering if adding return code 0x1 to ObjectNode.cpp/HandleSystemFailures as a system error, would be a good idea in your opinion? (juste where it says "// If remote PC is shutdown by user, compiler can be terminated"). You already seem to do this for clang and gcc.

**Comment by ffulin** [comment]
Thanks for reporting this. I was able to reproduce the problem.

I've made return code 0x1 now be considered a system failure (so the item will be retried). Some additional logic was needed, because Clang and GCC return 0x1 when terminating normally if there is a compile error (and we want to continue report that error rather than silently retrying).

I've made the necessary changes to fix this problem (https://github.com/fastbuild/fastbuild/commit/7f4189260f676787679f8f9fbb1258fb04b2e6de) and manually terminating tasks on the remote worker (both clang and cl) no longer causes the build to fail (and clang compile errors generated remotely are still correctly reported).

This fix will be in the next release (v0.96). I'll leave this issue open until that version is available.

**Comment by aganea** [comment]
I did the following change, similar to yours. Please note the additional changes to Client.cpp to prevent killing the <LOCAL RACE>, if any. In that case the local compilation should finish its job.

```
From e7cce35655f8a9d3772b476a190fc88bcc85b91d Mon Sep 17 00:00:00 2001
From: Alexandre Ganea <alexandre.ganea@ubisoft.com>
Date: Fri, 13 Apr 2018 17:01:39 -0400
Subject: [PATCH] Retry fix for remote builds that fail when the remote user
 kills the compilation

When a compilation process (cl.exe or orbis-clang++.exe) is spawned by the
FBuildWorker, sometimes that can make the machine unresponsible, and as a result
it can happen that the users are killing the compilation process manually.

When that happens the process terminates with code 1, as documented here:
https://stackoverflow.com/questions/4344923/process-exit-code-when-process-is-killed-forcibly
In that case the remote worker (FBuildWorker) will tag that job as a
"system failure", which will force the client to re-spawn the job on a different
worker. The client will also "blacklist" that worker as a result, which prevents
from spawning subsequent jobs on that client (as long as the client runs).

The change is a bit tricker when clang is used, because clang uses 1 as a exit
code when the compilation fails. We consequently check if the log contains
"error: ", and if not we consider the exit code as a call to TerminateProcess()
initiated by the remote user.
---
 Code/Tools/FBuild/FBuildCore/Graph/ObjectNode.cpp    | 16 ++++++++++++++++
 Code/Tools/FBuild/FBuildCore/Protocol/Client.cpp     | 12 ++++++++----
 Code/Tools/FBuild/FBuildCore/WorkerPool/JobQueue.cpp | 11 ++++++++++-
 Code/Tools/FBuild/FBuildCore/WorkerPool/JobQueue.h   |  2 +-
 4 files changed, 35 insertions(+), 6 deletions(-)

diff --git a/Code/Tools/FBuild/FBuildCore/Graph/ObjectNode.cpp b/Code/Tools/FBuild/FBuildCore/Graph/ObjectNode.cpp
index d66e36f..310a82b 100644
--- a/Code/Tools/FBuild/FBuildCore/Graph/ObjectNode.cpp
+++ b/Code/Tools/FBuild/FBuildCore/Graph/ObjectNode.cpp
@@ -2229,6 +2229,8 @@ bool ObjectNode::CompileHelper::SpawnCompiler( Job * job,
 //------------------------------------------------------------------------------
 /*static*/ void ObjectNode::HandleSystemFailures( Job * job, int result, const char * stdOut, const char * stdErr )
 {
+    UNUSED_VAR(stdErr);
+
     // Only remote compilation has special cases. We don't touch local failures.
     if ( job->IsLocal() )
     {
@@ -2262,6 +2264,9 @@ bool ObjectNode::CompileHelper::SpawnCompiler( Job * job,
         // MSVC errors
         if ( objectNode->IsMSVC() )
         {
+            // The compiler executable was manually killed by the user, through the TaskManager
+            // In that case the error code is set to 1 by TerminateProcess()
+            // See https://stackoverflow.com/questions/4344923/process-exit-code-when-process-is-killed-forcibly
             if ( result == 0x01 )
             {
                 job->OnSystemError();
@@ -2315,6 +2320,17 @@ bool ObjectNode::CompileHelper::SpawnCompiler( Job * job,
                     job->OnSystemError();
                     return;
                 }
+
+                // If exit code is 1, the task was probably killed manually by a remote user from the Task Manager.
+                // As clang already returns 1 as an error, we must check if the log contains the word "error: "
+                // If not, we assume a remote user killed the process manually.
+                // This precise error string is taken from the clang codebase, please see:
+                // https://github.com/llvm-mirror/clang/blob/a38ba9770a70056f9fdd6c71f819e5db45a105e4/tools/libclang/CIndexDiagnostic.cpp#L308
+                if ( !stdErr || strstr(stdErr, "error: ") == nullptr )
+                {
+                    job->OnSystemError();
+                    return;
+                }
             }
         }
 
diff --git a/Code/Tools/FBuild/FBuildCore/Protocol/Client.cpp b/Code/Tools/FBuild/FBuildCore/Protocol/Client.cpp
index 82b86a6..1afdc51 100644
--- a/Code/Tools/FBuild/FBuildCore/Protocol/Client.cpp
+++ b/Code/Tools/FBuild/FBuildCore/Protocol/Client.cpp
@@ -545,9 +545,16 @@ void Client::Process( const ConnectionInfo * connection, const Protocol::MsgJobR
         VERIFY( ss->m_Jobs.FindDerefAndErase( jobId ) );
     }
 
+    // was it a system error?
+    if (systemError)
+    {
+        // blacklist misbehaving worker
+        ss->m_Blacklisted = true;
+    }
+
     // Has the job been cancelled in the interim?
     // (Due to a Race by the main thread for example)
-    Job * job = JobQueue::Get().OnReturnRemoteJob( jobId );
+    Job * job = JobQueue::Get().OnReturnRemoteJob( jobId, systemError );
     if ( job == nullptr )
     {
         // don't save result as we were cancelled
@@ -626,9 +633,6 @@ void Client::Process( const ConnectionInfo * connection, const Protocol::MsgJobR
         // was it a system error?
         if ( systemError )
         {
-            // blacklist misbehaving worker
-            ss->m_Blacklisted = true;
-
             // take note of failure of job
             job->OnSystemError();
 
diff --git a/Code/Tools/FBuild/FBuildCore/WorkerPool/JobQueue.cpp b/Code/Tools/FBuild/FBuildCore/WorkerPool/JobQueue.cpp
index 6a08720..02cacd0 100644
--- a/Code/Tools/FBuild/FBuildCore/WorkerPool/JobQueue.cpp
+++ b/Code/Tools/FBuild/FBuildCore/WorkerPool/JobQueue.cpp
@@ -335,7 +335,7 @@ Job * JobQueue::GetDistributableJobToRace()
 
 // OnReturnRemoteJob
 //------------------------------------------------------------------------------
-Job * JobQueue::OnReturnRemoteJob( uint32_t jobId )
+Job * JobQueue::OnReturnRemoteJob( uint32_t jobId, const bool remoteJobFailed )
 {
     MutexHolder m( m_DistributedJobsMutex );
     auto jobIt = m_DistributableJobs_InProgress.FindDeref( jobId );
@@ -364,6 +364,15 @@ Job * JobQueue::OnReturnRemoteJob( uint32_t jobId )
         // Are we still locally racing?
         if ( distState == Job::DIST_RACING )
         {
+            if ( remoteJobFailed )
+            {
+                // take note of failure of job
+                job->OnSystemError();
+
+                // Let the local job finish, because the remote job failed (because of a system error)
+                return nullptr;
+            }
+
             // Try to cancel the local job
             job->Cancel();
             job->SetDistributionState( Job::DIST_RACE_WON_REMOTELY_CANCEL_LOCAL );
diff --git a/Code/Tools/FBuild/FBuildCore/WorkerPool/JobQueue.h b/Code/Tools/FBuild/FBuildCore/WorkerPool/JobQueue.h
index faf6c1c..bb48a86 100644
--- a/Code/Tools/FBuild/FBuildCore/WorkerPool/JobQueue.h
+++ b/Code/Tools/FBuild/FBuildCore/WorkerPool/JobQueue.h
@@ -80,7 +80,7 @@ private:
     // client side of protocol consumes jobs via this interface
     friend class Client;
     Job *       GetDistributableJobToProcess( bool remote );
-    Job *       OnReturnRemoteJob( uint32_t jobId );
+    Job *       OnReturnRemoteJob( uint32_t jobId, const bool remoteJobFailed = false );
     void        ReturnUnfinishedDistributableJob( Job * job );
 
     // Semaphore to manage work
-- 
2.14.1.windows.1


```

**Comment by ffulin** [comment]
A version containing a fix for this issue has now been released (v0.96).
